// @starport/vue/src/utils/nft.ts
export interface idCollection{
  denom_id: string;
  token_ids: string[];
}

export interface ownerNft{
  address: string;
  id_collections: idCollection[];
}

export interface pages{
  next_key: string;
  total: string;
}

export interface queryNftOwnerResponse{
  owner: ownerNft;
  pagination: pages;
}

/**
 * param struct for issue denom tx
 */
export interface IssueDenomParam {
  id:string;
  name:string;
  schema:string;
  sender:string;
}

/**
 * param struct for Mint NFT
 */
export interface MintNFTParam {
  id:string;
  denom_id:string;
  name:string;
  uri:string;
  data:string;
  sender:string;
  recipient:string;
}

/**
 * param struct for Edit NFT tx
 */
export interface EditNFTParam {
  id:string;
  denom_id:string;
  name?:string;
  uri?:string;
  data?:string;
  sender:string;
}

/**
 * param struct for Transfer NFT tx
 */
export interface TransferNFTParam {
  id:string;
  denom_id:string;
  name?:string;
  uri?:string;
  data?:string;
  sender:string;
  recipient:string;
}

/**
 * param struct for Burn NFT tx
 */
export interface BurnNFTParam {
  id:string;
  denom_id:string;
  sender:string;
}

export interface Denom {
    id:string;
    name:string;
    schema:string;
    creator:string;
}
